### ISSUE TITLE
how to efficiently propagate messages more times in one graphconv layer

### ISSUE THREAD

**Issue by scottjiao** [dmlc/dgl#2396]
Dear contributors,

The graphconv layer has implemented $AXW$. Now I want to implement the layer of IGCN (Label Efficient Semi-Supervised Learning via Graph Filtering), where the message passing process can be simply rewritten as $A^k XW$. 

I found that I can not simply implement this layer with GraphConv. My first solution is to build $k$ GraphConv layers without weight and bias as the convolution, and then build a nn.Linear layer as the weight, pack them in a order into a Sequential object renamed as GraphConv and put it into my model. 

However this method will cause a terrible time complexity propotional to $k$ for the whole training. As $k$ increases up to 10~30 the cost will be intoleroble. Hence I am writiting this issue. Do you have any suggestions on this seemly simple problem?

Best Regards.

**Comment by yzh119** [comment]
@scottjiao I think SGC did exactly the same thing, see https://github.com/dmlc/dgl/blob/master/python/dgl/nn/pytorch/conv/sgconv.py#L187

> As $k$ increases up to 10~30 the cost will be intoleroble. 

Not sure if it's the case, if $X$ is fixed then you can cache the result of $A^k X$ for training, and the overhead would be amortized. If not, you can either try pre-computing $A^k$ via [`khop_adj`](https://docs.dgl.ai/generated/dgl.khop_graph.html?highlight=khop#dgl.khop_adj) and directly computes the message passing via dense-dense matrix multiplication, or applies A * h one following another (A * (A * ... (A * h)) via dgl's message passing api (`update_all`), which one is faster is not sure (depending on the property of your graph).

But 10~30 is too high in my opinion... I suspect $A^k$ would be too large, most graphs' diameter is not that large.

**Comment by VoVAllen** [comment]
Also you can precompute A^k and store it as edata on the graph. And you change the message function from `copy_u` to `u_mul_e` to calculate the A^k X

**Comment by yzh119** [comment]
@VoVAllen I don't think so... how to store the weight of edge which does not exist in the original graph?

**Comment by VoVAllen** [comment]
I see. Please ignore my answer

**Comment by scottjiao** [comment]
> @scottjiao I think SGC did exactly the same thing, see https://github.com/dmlc/dgl/blob/master/python/dgl/nn/pytorch/conv/sgconv.py#L187
> 
> > As $k$ increases up to 10~30 the cost will be intoleroble.
> 
> Not sure if it's the case, if $X$ is fixed then you can cache the result of $A^k X$ for training, and the overhead would be amortized. If not, you can either try pre-computing $A^k$ via [`khop_adj`](https://docs.dgl.ai/generated/dgl.khop_graph.html?highlight=khop#dgl.khop_adj) and directly computes the message passing via dense-dense matrix multiplication, or applies A * h one following another (A * (A * ... (A * h)) via dgl's message passing api (`update_all`), which one is faster is not sure (depending on the property of your graph).
> 
> But 10~30 is too high in my opinion... I suspect $A^k$ would be too large, most graphs' diameter is not that large.

To some extent, the SGC layer solved my problem , thanks!
